Feature/svd qna thread

Dashadower opened the pull request:
I'll comment on some bits that's not working out atm. @bbbales2

bbbales2 commented on stan/math/rev/fun/singular_values_U.hpp:
```diff
@@ -0,0 +1,74 @@
+template <typename EigMat,
+    require_eigen_matrix_dynamic_t<EigMat>* = nullptr,
+	  require_vt_var<EigMat>* = nullptr>
+inline auto singular_values_U(const EigMat& m) {
```
Just make it `svd_U` (this sounds like it's the singular values of U)

bbbales2 commented on stan/math/rev/fun/singular_values_U.hpp:
```diff
@@ -0,0 +1,74 @@
+
+  auto arena_Fp = to_arena(Fp);
+  //auto arena_U = to_arena(svd.matrixU());
+  arena_t<ret_type> arena_U = (svd.matrixU());
```
This is the correct line.

`arena_t<...>` means that `arena_U` will live in the arena

The `ret_type` is an `Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>` here, and will make sure and promote the `Eigen::MatrixXd` output of `svd.matrixU()` to an autodiff type so it has adjoints and stuff (`U` needs to be an autodiff type cause we're gonna return it).

bbbales2 reviewed (commented):
Compiles now, but isn't working. I may have introduced bugs!

I merged in develop so that `.val_op()` would work. Sometimes we gotta use `.val_op()` instead of `.val()`. I always forget the rules lol and just switch when I get these big errors.

Dashadower commented:
Thanks. I'll get it finished within the weekend.

bbbales2 commented on stan/math/prim/fun/svd_U.hpp:
```diff
@@ -0,0 +1,32 @@
+	  require_eigen_matrix_dynamic_t<EigMat>* = nullptr,
+	  require_not_vt_var<EigMat>* = nullptr>
+Eigen::Matrix<value_type_t<EigMat>, Eigen::Dynamic, Eigen::Dynamic>
+svd_U(const EigMat& m) {
```
This prim implementation handles all the non-reverse mode autodiff stuff and the non-autodiff stuff.

So in terms of scalar types, that is `double`, `fvar<double>`, `fvar<var>`, `fvar<fvar<double>>` and `fvar<fvar<var>>`

bbbales2 commented on stan/math/rev/fun/svd_U.hpp:
```diff
@@ -0,0 +1,74 @@
+  
+  Eigen::JacobiSVD<Eigen::MatrixXd> svd(arena_m.val(), Eigen::ComputeFullU | Eigen::ComputeFullV);
+
+  auto arena_D = to_arena((Eigen::MatrixXd) svd.singularValues().asDiagonal());
```
This matrix isn't necessarily square, and if you uncomment the failing mixed mode tests I think this is causing a problem.

If the input is nonsquare, then this matrix has some extra zero rows or columns. You don't actually need to allocate a matrix with zero rows and columns --  you can just keep that in mine when you do operations with the matrix.

Check out how the autodiff in `singular_values` works. Instead of appending zeros to `D`, to compute `D * V^T`, we ignore the part of the matrix `V^T` that we'd multiply the zero stuff by.

bbbales2 commented on test/unit/math/mix/fun/svd_U_test.cpp:
```diff
@@ -0,0 +1,30 @@
+  m23 << 3, 5, -7, -11, 13, -17;
+  Eigen::MatrixXd m32 = m23.transpose();
+  stan::test::expect_ad(tols, f, m23);
+  stan::test::expect_ad(tols, f, m32);*/
```
These tests were segfaulting. Uncomment to get them workin'!

bbbales2 commented on test/unit/math/mix/fun/svd_U_test.cpp:
```diff
@@ -0,0 +1,30 @@
+
+  stan::test::ad_tolerances tols;
+  //tols.hessian_hessian_ = 1e-2;
+  //tols.hessian_fvar_hessian_ = 1e-2;
```
These options were used to limit the tolerance of the finite difference tests. If the tests work without them, nothing to worry about.

bbbales2 commented on test/unit/math/rev/fun/singular_values_U_test.cpp:
```diff
@@ -0,0 +1,47 @@
+#include <gtest/gtest.h>
+#include <test/unit/util.hpp>
+#include <typeinfo>
+TEST(AgradRev, svd_U_gradient) {
```
We don't actually need tests in rev. The mixed mode tests will cover this. We do need tests in prim for `svd_U` though.

The assumption of `expect_ad` is that the prim version returns accurate values. Once that is true, it can finite difference and check all the autodiff.

You'll need to implement the `prim` checks though. See how they're done for `eigenvectors_sym` or `singular_values`. You can just compute some reference values in R or something to compare things against.

bbbales2 reviewed (commented):
Comments!

Dashadower commented:
Rev tests for svd_V and svd_U always yield adjoint values of 0.

t4c1 committed:
fix doxygen

rok-cesnovar committed:
test win gpu

rok-cesnovar committed:
revert the rest of jenkinsfile

rok-cesnovar committed:
remove copy_triangular

rok-cesnovar committed:
remove diagonal multiply

rok-cesnovar committed:
remove identity()

rok-cesnovar committed:
Merge pull request #2267 from bstatcomp/kg_size_check

added size match checks to kg binary functions

rok-cesnovar committed:
remove sub_block

bbbales2 committed:
Fixed log_softmax bug, sorta added trace_gen_quad_form (Issue #2101)

rok-cesnovar committed:
Merge pull request #2260 from bstatcomp/auto_benchmarks

Add automatic benchmarking utility

bbbales2 committed:
varmat version of trace_gen_quad_form added (Issue #2101)

rok-cesnovar committed:
add add_diag rev

rok-cesnovar committed:
added unary plus/minus
added mean

rok-cesnovar committed:
fix jenkinsfile

rok-cesnovar committed:
finish prim

rok-cesnovar committed:
change minus/plus to unary operator

rok-cesnovar committed:
add sub_col, sub_row

rok-cesnovar committed:
add segment

rok-cesnovar committed:
cpplint

bbbales2 commented on test/unit/math/prim/fun/svd_U_test.cpp:
```diff
@@ -0,0 +1,14 @@
+#include <gtest/gtest.h>
+#include <test/unit/util.hpp>
+
+TEST(MathMatrixPrimMat, svd_U) {
```
For the tests in prim, test a 0x0 matrix, a 1x1 matrix, a small square matrix (2x2), a small tall rectangular matrix (3x2), and a small wide rectangular matrix (2x3).

We need to be sure the prim function is working before we can trust the `expect_ad` stuff to work (since it depends on finite differences as reference)

bbbales2 commented on test/unit/math/prim/fun/svd_V_test.cpp:
```diff
@@ -0,0 +1,16 @@
+#include <gtest/gtest.h>
+#include <test/unit/util.hpp>
+
+TEST(MathMatrixPrimMat, svd_V) {
```
Same tests for svd_V as for U.

rok-cesnovar committed:
add prim beta

bbbales2 commented on stan/math/rev/fun/svd_U.hpp:
```diff
@@ -0,0 +1,77 @@
+  const int M = to_arena(std::min(m.rows(), m.cols()));
+  auto arena_m = to_arena(m); // N by P
+  
+  Eigen::JacobiSVD<Eigen::MatrixXd> svd(arena_m.val(), Eigen::ComputeFullU | Eigen::ComputeFullV);
```
Also let's switch from computing the full U and V to the thin U and V.

It's discussed here: https://eigen.tuxfamily.org/dox/classEigen_1_1JacobiSVD.html

It might mean that you don't have to take the subblocks of U and V in the gradients anymore.

Goodrich says we probly only need the thin version of this: https://github.com/stan-dev/math/issues/2192#issuecomment-749310928

bbbales2 commented on test/unit/math/mix/fun/svd_U_test.cpp:
```diff
@@ -0,0 +1,30 @@
+#include <test/unit/math/test_ad.hpp>
+
+TEST(MathMixMatFun, svd_U) {
```
These tests compile for me but they fail when they run. Must be a bug somewhere. Comment them out and introduce them one by one till you find it.

bbbales2 commented on test/unit/math/mix/fun/svd_V_test.cpp:
```diff
@@ -0,0 +1,30 @@
+#include <test/unit/math/test_ad.hpp>
+
+TEST(MathMixMatFun, svd_V) {
```
Same as the svd_U tests. Uncomment the tests one by one and see if you can find the problem

bbbales2 reviewed (changes requested):
Review!

rok-cesnovar committed:
add rev beta

rok-cesnovar committed:
Merge pull request #2219 from stan-dev/add_opencl_cpu_tests

avoid copies in matrix_cl & add OpenCL CPU tests

rok-cesnovar committed:
added rev cbrt

rok-cesnovar committed:
added rev ceil

rok-cesnovar committed:
added rev floor/round

rok-cesnovar committed:
added trunc

rok-cesnovar committed:
add diag_post_multiply, diag_pre_multiply

rok-cesnovar committed:
Merge branch 'develop' into opencl/rev-support-part-4

rok-cesnovar committed:
Merge branch 'develop' into opencl_cleanup

# Conflicts:
#	stan/math/opencl/prim.hpp
#	test/unit/math/opencl/sub_block_test.cpp

rok-cesnovar committed:
added erf and erfc

rok-cesnovar committed:
added exp, exp2, expm1

rok-cesnovar committed:
added inv, inv_sqrt, inv_square, inv_cloglog, inv_logit

rok-cesnovar committed:
added rev lbeta

rok-cesnovar committed:
added lgamma

rok-cesnovar committed:
Added

log
log10
log1m
log1p
log2
logit
log1m_exp
log1p_exp
log_inv_logit
log1m_inv_logit

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

rok-cesnovar committed:
cleanup cholesky

rok-cesnovar committed:
remove unused test

rok-cesnovar committed:
cleanup doxygen and includes

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

rok-cesnovar committed:
added doxygen for diag_post/pre_multiply

rok-cesnovar committed:
fix doxygen

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

rok-cesnovar committed:
fix test names

rok-cesnovar committed:
fix doygen

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

rok-cesnovar committed:
fix cholesky typo

Dashadower committed:
Update tests for prim/fun/{svd_U, svd_V}.hpp. Change rev/fun/{svd_U, svd_V} functions to return thinned matrices instead of full

rok-cesnovar committed:
Apply suggestions from code review

Co-authored-by: Tadej Ciglarič <tadej.c@gmail.com>

rok-cesnovar committed:
expand (l)beta tests

rok-cesnovar committed:
Apply suggestions from code review

Co-authored-by: Tadej Ciglarič <tadej.c@gmail.com>

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

t4c1 committed:
bugfix making a copy of expression when constructing assignment pair

yashikno committed:
Merge commit 'be482a212615b80319654108c9b1cf291de9a170' into HEAD

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

rok-cesnovar committed:
remove diag_{post, pre}_multiply

rok-cesnovar committed:
fix test

t4c1 committed:
tmp

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

t4c1 committed:
bugfix check_cl

t4c1 committed:
typo

t4c1 committed:
Merge branch 'opencl_dot_product' of https://github.com/bstatcomp/math into opencl_dot_product

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

t4c1 committed:
reimplemented using evaluators

t4c1 committed:
fix doxygen

rok-cesnovar committed:
fix lbeta test

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

rok-cesnovar committed:
add missing require

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

rok-cesnovar committed:
Merge pull request #2272 from stan-dev/opencl/cleanup

OpenCL: cleanup and rev support for add_diag, unary +-, mean, sub_col, sub_row, segment

rok-cesnovar committed:
Merge branch 'develop' into opencl/rev-support-part-4

rok-cesnovar committed:
Update stan/math/opencl/rev/log1p.hpp

Co-authored-by: Tadej Ciglarič <tadej.c@gmail.com>

rok-cesnovar committed:
address review comments

yashikno committed:
Merge commit '60671039ef3978074ef9ba5420c5bda9757254cb' into HEAD

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

rok-cesnovar committed:
Merge pull request #2273 from stan-dev/opencl/rev-support-part-4

OpenCL: add rev support for (l)beta, cbrt, ceil, floor, round, trunc

rok-cesnovar committed:
Merge branch 'develop' into opencl/log-rev

rok-cesnovar committed:
Merge branch 'develop' into opencl/unary-rev

# Conflicts:
#	stan/math/opencl/rev.hpp

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

rok-cesnovar committed:
remove unnecessary requires

t4c1 committed:
fix cache test

t4c1 committed:
Merge branch 'opencl_dot_product' of https://github.com/bstatcomp/math into opencl_dot_product

t4c1 committed:
Merge branch 'develop' into opencl_dot_product

# Conflicts:
#	stan/math/opencl/rev.hpp

rok-cesnovar committed:
add missing require to sqrt

rok-cesnovar committed:
add elt_multiply/elt_divide includes

rok-cesnovar committed:
fix lgamma test

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

rok-cesnovar committed:
Merge pull request #2274 from stan-dev/opencl/log-rev

OpenCL: Add rev support for unary log* functions

rok-cesnovar committed:
Merge branch 'develop' into opencl/unary-rev

rok-cesnovar committed:
fix lgamma test

rok-cesnovar committed:
Merge pull request #2275 from stan-dev/opencl/unary-rev

OpenCL: rev support for erf, erfc, exp, exp2, epm1, inv, inv_sqrt, inv_square, inv_cloglog, inv_logit, lgamma

t4c1 committed:
bugfix

t4c1 committed:
Merge branch 'develop' into opencl_dot_product

# Conflicts:
#	stan/math/opencl/rev.hpp

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

t4c1 committed:
adds option to benchmark opencl function including copying time only for autodiff variables.

t4c1 committed:
bugfix it another way

t4c1 committed:
Merge branch 'opencl_dot_product' of https://github.com/bstatcomp/math into opencl_dot_product

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

t4c1 committed:
added copying results back to opencl copy benchmarks

bbbales2 committed:
Merge remote-tracking branch 'origin/develop' into feature/varmat-constrains

bbbales2 committed:
Don't use noalias and do use make callback var (Issue #2101)

bbbales2 committed:
Cleaned up `multiply_log` (Issue #2101)

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

bbbales2 committed:
Edits to constrain functions (Issue #2101)

bbbales2 committed:
Merge branch 'feature/varmat-constrains' of github.com:stan-dev/math into feature/varmat-constrains

bbbales2 committed:
Updated indexing in constrain functions (Issue #2101)

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

SteveBronder committed:
adds specializations for var_value<Matrix> for array and column vector conversions and adds helpers for extracting values for each

bbbales2 committed:
Merge remote-tracking branch 'origin/develop' into feature/varmat-three-argument-tests

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

SteveBronder committed:
fix headers

SteveBronder committed:
fix headers in distributions

SteveBronder committed:
update headers

Dashadower committed:
Update prim tests

Dashadower commented:
@bbales All tests passing, no more segfaults!

rok-cesnovar committed:
fix is_constant for kernel expressions

SteveBronder committed:
fix more headers

bbbales2 committed:
Broke out matvar tests into a separate file. Added three argument tests and tests for the tests (Issue #2101)

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

bbbales2 committed:
Make sure inputs to wishart/inv_wishart functions are symmetric (Issue #2101)

rok-cesnovar committed:
remove floating point restriction for isfinite

bbbales2 committed:
Updated wishart and inv_wishart functions to check symmetry + positive-definiteness on input arguments (and use the whole input matrix too) (Issue #2101)

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

bbbales2 committed:
Fixed fabs gradients and added default tolerance option to expect_ad_vector_matvar (Issue #2101)

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

SteveBronder committed:
update as_value_* functions to handle rvalues

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

bbbales2 committed:
Don't test abs gradients at 0 cause they don't exist (Issue #2101)

bbbales2 committed:
Merge branch 'feature/varmat-three-argument-tests' of github.com:stan-dev/math into feature/varmat-three-argument-tests

SteveBronder committed:
Merge pull request #2254 from stan-dev/feature/varmat-constrains

Initial commit to make matrix constrains varmat friendly (Issue #2101)

rok-cesnovar committed:
binomial_lpmf fixes

rok-cesnovar committed:
fixes for double_exponential_lpdf

bbbales2 committed:
Merge remote-tracking branch 'origin/develop' into feature/varmat-three-argument-tests

bbbales2 committed:
Merge pull request #2281 from stan-dev/bugfix/issue-2279

Make `wishart` and `inv_wishart` functions check symmetry/positive definiteness on input

rok-cesnovar committed:
fixs for logistic_lpdf

wds15 committed:
upgrading to sundials v5.6.1; removing old sundials library

wds15 committed:
upgrading to sundials v5.6.1; modifying with new version number

wds15 committed:
upgrading to sundials v5.6.1; adding unmodified sundials library

wds15 committed:
upgrading to sundials v5.6.1; pruning files

wds15 committed:
upgrading to sundials v5.6.1; adding config files

wds15 committed:
upgrading to sundials v5.6.1; removing printf and fprintf for CRAN

bbbales2 reviewed (approved):
This looks really good! Can you add a 3x2 to all the tests everywhere in prim and mix? After that make a pull request into Stan math with this (and do any the final cleanup there).

Edit: And when you make the pull it will ask for an issue, use this one: https://github.com/stan-dev/math/issues/2192

bbbales2 committed:
Merge remote-tracking branch 'origin/develop' into feature/1710-von-mises-cdf

rok-cesnovar committed:
Merge pull request #2283 from stan-dev/feature/issue-update-sundials-561

Feature/issue update sundials 561

rok-cesnovar committed:
fix for neg_binomial_2

rok-cesnovar committed:
fix for calc_if

Dashadower committed:
Update mix mode tests

yashikno committed:
Merge commit 'be49360ca7eb65addfa7001c04699fd6f1a85330' into HEAD

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

Dashadower committed:
Fix lint warnings

Dashadower committed:
Update doxygen warnings

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

Dashadower committed:
Update doxygen docs

rok-cesnovar committed:
fix neg_binomial_2_log

SteveBronder committed:
fixup multiply

SteveBronder committed:
update names

t4c1 committed:
Merge pull request #2247 from bstatcomp/multi_expression2

Multi expression

yashikno committed:
Merge commit 'b2eaa329ee3427245dbe06482ee937569cb6032f' into HEAD

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

SteveBronder committed:
Fix for smooth access in elt_multiply

SteveBronder committed:
Fix for smooth access in elt_multiply

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

bbbales2 committed:
Merge remote-tracking branch 'origin/develop' into feature/varmat-three-argument-tests

bbbales2 committed:
Added test for callback_vari change and changed how temporaries in `trace_gen_quad_form` work (Issue #2101)

SteveBronder committed:
Merge remote-tracking branch 'origin/develop' into feature/varmat-array_or_scalar

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

bbbales2 committed:
Changed naming in test_ad_matvar and updated test_ad_matvar_test tests to return autodiff types if any of the inputs are autodiff types (Issue #2101)

bbbales2 committed:
Merge branch 'feature/varmat-three-argument-tests' of github.com:stan-dev/math into feature/varmat-three-argument-tests

bbbales2 committed:
Added `for_each` and used it in `expect_ad_matvar` tests (Issue #2101)

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

SteveBronder committed:
remove all the specializations and just use value_of() and coercing function. Fixup tests to check for types and move var versions over to rev

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

t4c1 committed:
exclude aggregates

t4c1 committed:
move definition of check_cl_::operator= to multi_result_kernel.hpp

t4c1 committed:
Merge branch 'develop' into opencl_dot_product

# Conflicts:
#	stan/math/prim/fun/square.hpp

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

t4c1 committed:
Merge branch 'benchmarking_fixes' into benchmark_opencl_copy_rev

Dashadower committed:
Merge branch 'develop' of https://github.com/stan-dev/math into feature/svd_review

Dashadower committed:
Update tests and functions for svd

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

bbbales2 committed:
Merge pull request #2280 from stan-dev/feature/varmat-three-argument-tests

Three argument tests for varmat (and varmat implementation of `trace_gen_quad_form`)

SteveBronder committed:
remove unneeded loop in elt_multiply

rok-cesnovar committed:
Merge branch 'develop' into opencl/minor_fixes

rok-cesnovar committed:
added test for beta_lpdf

rok-cesnovar committed:
add test for n, N scalar

rok-cesnovar committed:
add test for double_exponential_lpdf

rok-cesnovar committed:
add test for logistic_lpdf

bbbales2 committed:
Updated NaN check for von_mises_cdf (Issue #1710)

SteveBronder committed:
fix opencl multiply docs

SteveBronder committed:
update multiply to use to_arena()

yashikno committed:
Merge commit '7d5f626b55d822f2544138cc5fbeabbd4d86a92c' into HEAD

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

SteveBronder committed:
adds val() method for seq_scalar_view and vector_seq_view

rok-cesnovar committed:
test for neg_binomial_2

SteveBronder committed:
add tests for var scalar and vector seq views

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

SteveBronder committed:
fix header guards

rok-cesnovar committed:
added test for calc_if(double)

rok-cesnovar committed:
add test for neg_bin_2_log

rok-cesnovar committed:
add test that classification function works with matrix_cl<int>

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

t4c1 committed:
fixed requires

yashikno committed:
Merge commit '7d5f626b55d822f2544138cc5fbeabbd4d86a92c' into HEAD

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

SteveBronder committed:
add data() member for pointer access

rok-cesnovar committed:
Apply suggestions from code review

Co-authored-by: Tadej Ciglarič <tadej.c@gmail.com>

rok-cesnovar committed:
Revert "added test for calc_if(double)"

This reverts commit 7dadd10a6e3d152405070282224b461ae3ebca35.

# Conflicts:
#	test/unit/math/opencl/kernel_generator/calc_if_test.cpp

bbbales2 committed:
Merge pull request #1753 from pgree/feature/1710-von-mises-cdf

Feature/1710 von mises cdf

rok-cesnovar committed:
Merge pull request #2270 from bstatcomp/opencl_dot_product

Implement OpenCL dot product related functions

rok-cesnovar committed:
Merge pull request #2293 from stan-dev/opencl/minor_fixes

OpenCL: minor fixes for distributions

SteveBronder committed:
fix c_ to t_ in scalar version of scalar sequence view

SteveBronder committed:
add data() member to seq_scalar_view

yashikno committed:
Merge commit 'a280e0d5bc5556c5a3fa5f100e1b8931d61f8f5c' into HEAD

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

rok-cesnovar committed:
added sqrt and square

rok-cesnovar committed:
added tgamma

rok-cesnovar committed:
added Phi

rok-cesnovar committed:
added Phi_approx

rok-cesnovar committed:
added inv_Phi

SteveBronder committed:
use decltype(auto) for val() *_seq_view classes

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

SteveBronder committed:
update tests to move from meta to fun in prim for seq views

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

SteveBronder committed:
fix glm for accessing pointer to data

SteveBronder committed:
Merge branch 'feature/varmat-seq-scalar' of github.com:stan-dev/math into feature/varmat-seq-scalar

SteveBronder committed:
fix beta dist

SteveBronder committed:
update student and pareto

SteveBronder committed:
fix binomial

rok-cesnovar committed:
added fabs

rok-cesnovar committed:
added ldexp

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

rok-cesnovar committed:
fix fabs_test

bbbales2 committed:
Merge pull request #2278 from stan-dev/feature/varmat-array_or_scalar

Add var<mat> specializations for array and column vector converters

bbbales2 committed:
Merge remote-tracking branch 'origin/develop' into feature/varmat-seq-scalar

bbbales2 committed:
Added size tests to vector sequence view (Issue #2101)

rok-cesnovar committed:
uncomment test

yashikno committed:
Merge commit 'aa1c1c48372fb8e60da9ca7af6f0403b83f99d9d' into HEAD

rok-cesnovar committed:
test from [0,1]

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

rok-cesnovar committed:
Merge pull request #2294 from stan-dev/opencl/unary-simple-and-ldexp

OpenCL: add rev support for sqrt, square, tgamma, Phi, Phi_approx, inv_Phi, fabs, ldexp

bbbales2 committed:
Merge pull request #2290 from bstatcomp/benchmark_opencl_copy_rev

Benchmark improvements

SteveBronder committed:
ignores .cpp and executable files in benchmarks, changes permissions of benchmarks.py, adds BENCHMARK_* flags for make/local, and removes if catch for elt_multiply

yashikno committed:
Merge commit 'e25cd489144226e47fbb17df45051452e7985847' into HEAD

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)

SteveBronder committed:
Merge pull request #2292 from stan-dev/feature/varmat-seq-scalar

Make sequence views compatible with var<matrix>

Dashadower committed:
Merge branch 'develop' of https://github.com/stan-dev/math into feature/svd_review

Dashadower committed:
Add vari support and relevant tests

bbbales2 committed:
Merge pull request #2289 from stan-dev/cleanup/multiply-varmat

Cleanup/multiply varmat

bbbales2 committed:
Removed extra comments, rearranged includes, and removed custom tolerances in tests (Issue #2101)

yashikno committed:
Merge commit '88040ea5e8e8bc308ee080a8dcab9d5452f40ecf' into HEAD

stan-buildbot committed:
[Jenkins] auto-formatting by clang-format version 6.0.0-1ubuntu2~16.04.1 (tags/RELEASE_600/final)
